Adds microbenchmarking and uses standard 1e9 for memory calcs.

patemotter opened the pull request:
Will need to add a smoke test appended to this pr or in the next pr when adding/changing the benchmark.

rwitten commented on MaxText/decode.py:
```diff
@@ -23,6 +23,9 @@
 import pyconfig
 import sys
 
+from jax.experimental.compilation_cache import compilation_cache as cc
```
Can you put this into pyconfig as we discussed?

rwitten commented on MaxText/microbenchmark.py:
```diff
@@ -0,0 +1,184 @@
+"""Tests for a mock version of the engine API, used in integration tests elsewhere.
```
inference_microbenchmark? I think?

rwitten replied:
Also this prefix will cause problems (and is wrong).

rwitten commented on MaxText/microbenchmark.py:
```diff
@@ -0,0 +1,184 @@
+    prefill_results.update(ar_results)
+    results[prefill_length] = prefill_results
+
+  with open("benchmark_results.json", "w") as f:
```
can you make sure that sharegpt analysis can consume this? (And you can remove idealization from the sharegpt analysis.)

rwitten commented on MaxText/microbenchmark.py:
```diff
@@ -0,0 +1,184 @@
+  jax.config.update('jax_default_prng_impl', 'unsafe_rbg')
+  os.environ["TF_CPP_MIN_LOG_LEVEL"] = "0"
+  os.environ["M_ICI_FSDP_PARALLELISM"] = "1"
+  os.environ["M_ICI_AUTOREGRESSIVE_PARALLELISM"] = "-1"
```
pretty intense. I'd like you to hold off on this. In the future we might want different settings.

rwitten commented on run_microbenchmark.sh:
```diff
@@ -0,0 +1,21 @@
+# export TPU_MIN_LOG_LEVEL=0
+# export TF_CPP_MIN_LOG_LEVEL=0
+export M_ICI_FSDP_PARALLELISM=1
+export M_ICI_AUTOREGRESSIVE_PARALLELISM=-1
```
can we put this into the inference scratch world?

rwitten reviewed (changes requested):
Some nits!

P.S. - for faster responses in the future, please mark me as having next steps!

patemotter commented:
1. Improved readability of benchmark output. 
2. Now prints two lines that you can either copy and paste into sharegpt_analyze.py or you can load the json output directly. 
3. Refactored the analyze_sharegpt.py script. 
4. Moved cache compilation method out of individual scripts and into pyconfig.

patemotter commented:
Removed unnecessary engine inits.
